examples/custom-environment/lib.rs

```diff
@@ -0,0 +1,126 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink::env::{
+    DefaultEnvironment,
+    Environment,
+};
+
+/// Our custom environment diverges from the `DefaultEnvironment` in the event topics limit.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum EnvironmentWithManyTopics {}
+
+impl Environment for EnvironmentWithManyTopics {
+    // We allow for 5 topics in the event, therefore the contract pallet's schedule must allow for
+    // 6 of them (to allow the implicit topic for the event signature).
+    const MAX_EVENT_TOPICS: usize =
+        <DefaultEnvironment as Environment>::MAX_EVENT_TOPICS + 1;
+
+    type AccountId = <DefaultEnvironment as Environment>::AccountId;
+    type Balance = <DefaultEnvironment as Environment>::Balance;
+    type Hash = <DefaultEnvironment as Environment>::Hash;
+    type BlockNumber = <DefaultEnvironment as Environment>::BlockNumber;
+    type Timestamp = <DefaultEnvironment as Environment>::Timestamp;
+
+    type ChainExtension = <DefaultEnvironment as Environment>::ChainExtension;
+}
+
+#[ink::contract(env = crate::EnvironmentWithManyTopics)]
+mod runtime_call {
+    /// Trivial contract with a single message that emits an event with many topics.
+    #[ink(storage)]
+    #[derive(Default)]
+    pub struct Topics;
+
+    /// An event that would be forbidden in the default environment, but is completely valid in
+    /// our custom one.
+    #[ink(event)]
+    #[derive(Default)]
+    pub struct EventWithTopics {
+        #[ink(topic)]
+        first_topic: Balance,
+        #[ink(topic)]
+        second_topic: Balance,
+        #[ink(topic)]
+        third_topic: Balance,
+        #[ink(topic)]
+        fourth_topic: Balance,
+        #[ink(topic)]
+        fifth_topic: Balance,
+    }
+
+    impl Topics {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Default::default()
+        }
+
+        /// Emit an event with many topics.
+        #[ink(message)]
+        pub fn trigger(&mut self) {
+            self.env().emit_event(EventWithTopics::default());
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        #[ink::test]
+        fn emits_event_with_many_topics() {
+            let mut contract = Topics::new();
+            contract.trigger();
+
+            let emitted_events = ink::env::test::recorded_events().collect::<Vec<_>>();
+            assert_eq!(emitted_events.len(), 1);
+
+            let _ = <<Topics as ink::reflect::ContractEventBase>::Type as scale::Decode>::decode(&mut &emitted_events[0].data[..])
+                .expect("encountered invalid contract event data buffer");
+        }
+    }
+
+    #[cfg(all(test, feature = "e2e-tests", feature = "permissive-node"))]
```

Comment:
I'd also be curious to know what happens if we try and run an E2E test with a custom env on a node that doesn't use one

Reply:
test added